### PR TITLE
K8SPSMDB-359: Sharded data test

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-359](https://jira.percona.com/browse/K8SPSMDB-359) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

This test checks if data was rebalanced among multiple shards.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| scheduled-backup | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| demand-backup-sharded | passed |
| data-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-526-f9fe0a50

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| storage | passed |
| smart-update | passed |
| self-healing | passed |
| version-service | passed |
| operator-self-healing | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | failed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| scheduled-backup | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |
| data-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| demand-backup | failed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Review by cap1984 — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | failed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| security-context | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-526-8f06298a

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| init-deploy | passed |
| service-per-pod | passed |
| limits | passed |
| liveness | passed |
| scaling | passed |
| smart-update | passed |
| security-context | passed |
| version-service | passed |
| rs-shard-migration | passed |
| users | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| rs-shard-migration | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-526-f2451141

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| limits | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-526-77bd9f2e

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| demand-backup-sharded | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | failed |
| security-context | passed |
| operator-self-healing | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-526-58ca9582

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| storage | passed |
| data-sharded | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Review by qJkee — DISMISSED** [review]

File: e2e-tests/data-sharded/data [hunk 1]
```diff
@@ -0,0 +1,2002 @@
+[
```
Comment (tplavcic) [inline e2e-tests/data-sharded/data:1]: No need to commit actual data, just create a `data.js` file like this:
```
use app;
var bulk = db.city.initializeUnorderedBulkOp();
for(var i=0; i<200000; i++){
   bulk.insert( { "name": "city-"+i, "zipcode": i });
}
bulk.execute();
```
and then load it with the client.

File: e2e-tests/data-sharded/run [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+sleep 2
+
+desc 'set chunk size to 1 MB'
+run_mongos \
+    "use config\n db.settings.save( { _id:\"chunksize\", value: 1 } )" \
+    "clusterAdmin:clusterAdmin123456@$cluster-mongos.$namespace"
+sleep 2
+
+desc 'write data'
+for i in {1..100}; do
```
Comment (tplavcic) [inline e2e-tests/data-sharded/run:48]: loop is not needed here if `js` file is created with actual number of rows

**Review by tplavcic — CHANGES_REQUESTED** [review]

File: e2e-tests/data-sharded/data (hunk 1 above)
Comment (blez) [reply e2e-tests/data-sharded/data:1]: The problem is that my script inserts data in bulk (2000 records per request). And is't already not fast operation.
If start to insert 200000 records one by one this test will be slower almost 2000 times.

File: e2e-tests/data-sharded/run [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        "use config\n db.chunks.count({\"ns\" : \"app.city\", \"shard\": \"$i\"})" \
+        "clusterAdmin:clusterAdmin123456@$cluster-mongos.$namespace" |
+        grep "switched to db config" -A 1 | grep -v "switched to db config")
+
+    if [[ "$out" -ne 0 ]]; then
+        ((shards = shards + 1))
+    fi
+done
+
+if [[ "$shards" -eq 0 ]]; then
```
Comment (hors) [inline e2e-tests/data-sharded/run:68]: maybe we need to confirm that we have data on all shards...

Comment (blez) [reply e2e-tests/data-sharded/run]: https://github.com/percona/percona-server-mongodb-operator/pull/526/commits/0ef77690fcb99f70a404dee504f3552e1c48a1ec

File: e2e-tests/data-sharded/run (hunk 2 above)
Comment (blez) [reply e2e-tests/data-sharded/run]: https://github.com/percona/percona-server-mongodb-operator/pull/526/commits/0ef77690fcb99f70a404dee504f3552e1c48a1ec

File: e2e-tests/data-sharded/data (hunk 1 above)
Comment (blez) [reply e2e-tests/data-sharded/data]: https://github.com/percona/percona-server-mongodb-operator/commit/0ef77690fcb99f70a404dee504f3552e1c48a1ec

**Review by tplavcic — CHANGES_REQUESTED** [review]
This is ok to confirm that data is now sharded, but maybe we should extend it to confirm that operator will throw an error if we try to remove a shard.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | failed |
| arbiter | passed |
| security-context | failed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| upgrade | failed |
| rs-shard-migration | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

**Comment by blez** [comment]
@tplavcic I think it should be separate test.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scheduled-backup | failed |
| arbiter | passed |
| scaling | passed |
| service-per-pod | passed |
| security-context | failed |
| demand-backup-sharded | failed |
| operator-self-healing | passed |
| liveness | passed |
| smart-update | passed |
| rs-shard-migration | passed |
| upgrade | failed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| upgrade-consistency | passed |
| one-pod | passed |
| storage | passed |
| demand-backup | passed |
| limits | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | failed |
| operator-self-healing | passed |
| scheduled-backup | failed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | failed |
| users | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| upgrade | failed |

**Review by tplavcic — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | failed |
| storage | passed |
| one-pod | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| upgrade | failed |

**Review by hors — CHANGES_REQUESTED** [review]
please check failed tests

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | failed |
| storage | passed |
| one-pod | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| upgrade | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| upgrade | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | failed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| storage | passed |
| smart-update | passed |
| self-healing | passed |
| version-service | passed |
| operator-self-healing | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| rs-shard-migration | passed |
| upgrade | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | failed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| storage | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| self-healing | passed |
| liveness | passed |
| operator-self-healing | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| storage | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |
| upgrade | failed |
| data-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-526-102695b7

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

**Review by tplavcic — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Pull request merged**